python/strings_udf/cpp/include/cudf/strings/udf/udf_string.cuh

```diff
@@ -0,0 +1,475 @@
+                                                 cudf::char_utf8 chr)
+{
+  return replace(pos, 0, count, chr);
+}
+
+__device__ inline udf_string udf_string::substr(cudf::size_type pos, cudf::size_type count) const
+{
+  if (pos < 0) { return udf_string{"", 0}; }
+  auto const spos = byte_offset(pos);
+  if (spos >= m_bytes) { return udf_string{"", 0}; }
+  auto const epos = count < 0 ? m_bytes : std::min(byte_offset(pos + count), m_bytes);
+  return udf_string{data() + spos, epos - spos};
+}
+
+// utility for replace()
+__device__ void udf_string::shift_bytes(cudf::size_type spos,
+                                        cudf::size_type epos,
+                                        cudf::size_type nbytes)
+{
+  if (nbytes < m_bytes) {
+    // shift bytes to the left [...wxyz] -> [wxyzxyz]
+    auto src = epos;
+    auto tgt = spos;
+    while (tgt < nbytes) {
+      m_data[tgt++] = m_data[src++];
+    }
+  } else if (nbytes > m_bytes) {
+    // shift bytes to the right [abcd...] -> [abcabcd]
+    auto src = m_bytes;
+    auto tgt = nbytes;
+    while (src > epos) {
+      m_data[--tgt] = m_data[--src];
+    }
+  }
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
+                                                  cudf::size_type count,
+                                                  char const* str,
+                                                  cudf::size_type in_bytes)
+{
+  if (pos < 0 || in_bytes < 0) { return *this; }
+  auto const spos = byte_offset(pos);
+  if (spos > m_bytes) { return *this; }
+  auto const epos = count < 0 ? m_bytes : std::min(byte_offset(pos + count), m_bytes);
+
+  // compute new size
+  auto const nbytes = m_bytes + in_bytes - (epos - spos);
+  if (nbytes > m_capacity) { reallocate(2 * nbytes); }
+
+  // move bytes -- make room for replacement
+  shift_bytes(spos + in_bytes, epos, nbytes);
+
+  // insert the replacement
+  memcpy(m_data + spos, str, in_bytes);
+
+  m_bytes         = nbytes;
+  m_data[m_bytes] = 0;
+  return *this;
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
+                                                  cudf::size_type count,
+                                                  char const* str)
+{
+  return replace(pos, count, str, detail::bytes_in_null_terminated_string(str));
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
+                                                  cudf::size_type count,
+                                                  cudf::string_view const in)
+{
+  return replace(pos, count, in.data(), in.size_bytes());
+}
+
+__device__ inline udf_string& udf_string::replace(cudf::size_type pos,
+                                                  cudf::size_type count,
+                                                  cudf::size_type chr_count,
+                                                  cudf::char_utf8 chr)
```

Comment:
This question applies to other methods that have overloads for `char *` and `cudf::char_utf8`: do you see any ways to share more code between those overloads? I don't see any obvious solutions since in general one supports `memcpy` while the other does not, but I assume you have thought about this more than I have so was wondering if there were any additional approaches worth considering that I'm not thinking about.

Reply:
I will consider it. Is this required to approve this PR?

Reply:
No, happy to do this in a future PR if you come up with ideas.

Reply:
Nice that looks good.